core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -633,11 +661,33 @@ public ShareAcquiredRecords acquire(
                     firstBatch.baseOffset(), lastBatch.lastOffset(), batchSize, maxFetchRecords);
             }
 
-            log.trace("Overlap exists with in-flight records. Acquire the records if available for"
-                + " the share partition: {}-{}", groupId, topicIdPartition);
-            List<AcquiredRecords> result = new ArrayList<>();
             // The acquired count is used to track the number of records acquired for the request.
             int acquiredCount = 0;
+            List<AcquiredRecords> result = new ArrayList<>();
+
+            // If baseOffset is less than the first key of submap, this means the fetch happened for a gap in the cachedState.
+            // Thus a new batch needs to be acquired for the gap.
+            if (baseOffset < subMap.firstKey()) {
+                // This is to check whether the fetched records are all part of the gap, or they overlap with the next
+                // inFlight batch in the cachedState
+                if (lastBatch.lastOffset() < (subMap.firstKey())) {
+                    // The entire request batch is part of the gap
+                    return acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
+                        firstBatch.baseOffset(), lastBatch.lastOffset(), batchSize, maxFetchRecords);
+                } else {
+                    result.add(new AcquiredRecords()
+                        .setFirstOffset(firstBatch.baseOffset())
+                        .setLastOffset(subMap.firstKey() - 1)
+                        .setDeliveryCount((short) 1));
+                    acquiredCount += (int) (subMap.firstKey() - firstBatch.baseOffset());
+                    acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
+                        firstBatch.baseOffset(), subMap.firstKey() - 1, batchSize, maxFetchRecords);
+                }
+            }
```

Comment:
Surprising to see the code outside of the below for loop. Why do you want to consider only first gap, can't there be more gaps i.e. persister returned start offset 10 and acked batches as 15-20, 30-40. Subsequent fetch returned batches from 10-50, then how are we acquiring batch from 21-29? Do we have a test case?